### PR TITLE
Replace Block.Ref with Block reference count

### PR DESCRIPTION
**Pull request by dnhatn**
Block.Ref was introduced to help memory management for Evaluator. However, we can replace it with the Block referencing count now.

### CODE REVIEW

**Review by ChrisHegarty — APPROVED** [review]
LGTM Thanks @dnhatn

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java [hunk 1]
```diff
@@ -42,12 +42,7 @@ public EvalOperator(BlockFactory blockFactory, ExpressionEvaluator evaluator) {
 
     @Override
     protected Page process(Page page) {
-        Block.Ref ref = evaluator.eval(page);
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java:45]: question: is it correct to say that, with this PR, `eval` always returns an _owned_ block? I.e. the returned block needs to be closed (is always treated like a floating `Block.Ref` previously was)?

If so, IMO we should update the javadoc for `ExpressionEvaluator.eval` and mention that the returned block must be closed by the caller.

(This is not always the case: `Page.getBlock` for instance returns essentially an un-owned reference to a block and we need to manually `incRef` to take ownership of a page's block.)

File: x-pack/plugin/esql/src/main/java/generated/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAvgDoubleEvaluator.java [hunk 2]
```diff
@@ -34,55 +34,51 @@ public String name() {
...
-          int first = v.getFirstValueIndex(p);
-          int end = first + valueCount;
-          for (int i = first; i < end; i++) {
-            double value = v.getDouble(i);
-            MvAvg.process(work, value);
-          }
-          double result = MvAvg.finish(work, valueCount);
-          builder.appendDouble(result);
+  public Block evalNullable(Block fieldVal) {
+    DoubleBlock v = (DoubleBlock) fieldVal;
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/generated/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAvgDoubleEvaluator.java:38]: question: I guess `fieldVal` was never a floating `Block.Ref` when passed in here? We don't seem to need to close this block, so it probably is always part of a page and released via `Page.releaseBlocks`, right?

File: x-pack/plugin/esql/src/main/java/generated/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAvgIntEvaluator.java [hunk 3]
```diff
@@ -35,117 +35,109 @@ public String name() {
...
-            builder.appendDouble(result);
-            continue;
-          }
-          int end = first + valueCount;
-          for (int i = first; i < end; i++) {
-            int value = v.getInt(i);
-            MvAvg.process(work, value);
-          }
-          double result = MvAvg.finish(work, valueCount);
+  public Block evalNullable(Block fieldVal) {
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/generated/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAvgIntEvaluator.java:38]: Similarly here (and the other `eval...` functions inherited): should we update the javadoc for the super, abstract function? The contract is that the return value is an owned block, now, which must be closed by the caller.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java [hunk 4]
```diff
@@ -231,26 +231,25 @@ public Block.Ref eval(Page page) {
...
                             }
                         }
-                        try (Block.Ref elseRef = elseVal.eval(limited)) {
-                            result.copyFrom(elseRef.block(), 0, 1);
+                        try (Block values = elseVal.eval(limited)) {
+                            result.copyFrom(values, 0, 1);
                         }
                     }
                 }
-                return Block.Ref.floating(result.build());
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java:253]: Neat, this looks nice and clean!

**Review by alex-spies — APPROVED** [review]
Awesome! Only minor remarks/questions from my side :)

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java (hunk 1 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java:45]: ++ I added some javadocs in https://github.com/elastic/elasticsearch/pull/102513/commits/d56fcf2d005944d29a7bdfd360dcea1a4df51844

File: x-pack/plugin/esql/src/main/java/generated/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAvgDoubleEvaluator.java (hunk 2 above)
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/generated/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAvgDoubleEvaluator.java:38]: That's correct. We should not release the input block.

File: x-pack/plugin/esql/src/main/java/generated/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAvgIntEvaluator.java (hunk 3 above)
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/generated/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAvgIntEvaluator.java:38]: see https://github.com/elastic/elasticsearch/pull/102513/commits/d6d79023a4b934b937b879008a585e775692e944

**Comment by dnhatn** [comment]
@ChrisHegarty @alex-spies Thanks for reviewing.

**Pull request merged**